### PR TITLE
Wip Qt6 port

### PR DESCRIPTION
**Pull request by gfgit**
Depends on:
- [lxqt-build-tools#77](https://github.com/lxqt/lxqt-build-tools/pull/77)
- [liblxqt#337](https://github.com/lxqt/liblxqt/pull/337)

### CODE REVIEW

**Comment by stefonarch** [comment]
> Mainly for testing if `liblxqt` port is good.

On Neon I started porting lxqt-about, and after some more changes to lxqt-build-tools compiling failed at 80% with `#include QDialog no such file...
`

**Comment by stefonarch** [comment]
It compiles fine now and daemon starts, but no keybinds and translations are shown in config dialog.

I doubt over all those lxqt2, are they meant to be removed later?
```
-- Installing: /usr/local/bin/lxqt2-globalkeysd
-- Installing: /usr/local/bin/lxqt2-config-globalkeyshortcuts
-- Installing: /usr/local/share/applications/lxqt2-config-globalkeyshortcuts.desktop
-- Installing: /etc/xdg/autostart/lxqt2-globalkeyshortcuts.desktop
```


And for the translation on weblate those `*.yaml` should remain without suffix "2"  I guess:
`autostart/translations/lxqt2-globalkeyshortcuts_hr.desktop.yaml`

**Comment by gfgit** [comment]
> I doubt over all those lxqt2, are they meant to be removed later?

I don't know, I just followed what was done in `lxqt-build-tools` prefixing everything with `lxqt2`. Surely I missed some places to rename.

I think this is done to allow parallel installation of old Qt5 version and new Qt6 version. It might be useful also for testing the port while it's still WIP.

However this `globalkeys` is used by many application so I think should maintain compatibility with both Qt5 and Qt6.
This means maybe the library will change name so both version can be installed, but daemon, UI, autostart will have only one version installed.

**Comment by stefonarch** [comment]
> I think this is done to allow parallel installation of old Qt5 version and new Qt6 version. It might be useful also for testing the port while it's still WIP.

Afaik there is no idea to support parallel installations, the changes were needed to allow the qt6 branches of the already ported components which can coexist, but for all others this isn't planned afaik. 

The default share_dir "lxqt2" [should be changed](https://github.com/lxqt/lxqt-build-tools/pull/77#pullrequestreview-1507184738) as it breaks all themes and wallpapers when upgrading.

https://github.com/lxqt/lxqt/discussions/2139#discussioncomment-7479362

**Comment by stefonarch** [comment]
No error message when starting under wayland, but with openbox:

screen_area_ven_12:31:28_

**Comment by gfgit** [comment]
> No error message when starting under wayland, but with openbox:

What do you mean by OpenBox on Wayland? Also how is this related to code?

Anyway I can remove `lxqt2` prefixes and rebase the PR. Only `lxqt2-build-tool` must remain unless it is changed inside build tools repository

**Comment by stefonarch** [comment]
I first started the daemon under plasma wayland and adding shortcuts was disabled, but it should show a "Couldn't register shortcuts" message. 
Now I'm using openbox session in neon VM, plasma was too unstable.

Last check out fails here:

```
[ 31%] Building CXX object daemon/CMakeFiles/lxqt2-globalkeysd.dir/core.cpp.o
/home/user/gfgit/lxqt-globalkeys/daemon/core.cpp:39:10: fatal error: std::pair: No such file or directory
   39 | #include <std::pair>
      |          ^~~~~~~~~~~
```

**Comment by stefonarch** [comment]
Changed to ` #include <utility>` in` daemon/core.cpp` fixed it here.

**Comment by tsujan** [comment]
> I think this is done to allow parallel installation of old Qt5 version and new Qt6 version

I'm afraid that's not the case. Except for lxqt-build-tools, libfm-qt, libqtxdg and lxqt-qtplugin, which can be installed in parallel for Qt5 and Qt6, other components should be overwrite their Qt5-based versions. The reason is simple: we want to support styling of Qt5 apps and have LXQt file dialog in them for a long time.

**Comment by tsujan** [comment]
I wasn't following the discussion, but the cause of troubles may have been what I warned about at https://github.com/lxqt/lxqt-build-tools/pull/77#discussion_r1247857714. Sadly, @luis-pereira hasn't paid attention to it, and I don't intend to make things more complicated by adding commits to his PR.

**Comment by gfgit** [comment]
Ok now it's more clear. I will wait for the change on build tools repository and then rebase my PRs

**Comment by stefonarch** [comment]
Didn't really investigate but the  translations of `*.ts` files are not generated.

**Comment by tsujan** [comment]
> Didn't really investigate but the translations of *.ts files are not generated.

Last sentence of https://github.com/lxqt/lxqt-build-tools/pull/77#discussion_r1247857714?

```
Therefore, for example, there's no reason to have different LXQT_TRANSLATIONS_DIR or LXQT_GRAPHICS_DIR folders.
```

**Comment by stefonarch** [comment]
> > Didn't really investigate but the translations of *.ts files are not generated.
> 
> Last sentence of [lxqt/lxqt-build-tools#77 (comment)](https://github.com/lxqt/lxqt-build-tools/pull/77#discussion_r1247857714)?


No, innocent here. In my trials to port lxqt-about and lxqt-runner translations are generated. But probably the "2" has to do with it.

**Comment by stefonarch** [comment]
Renaming also the '*.ts` files to prefix "lxqt2-" :
`[ 22%] Generating lxqt2-config-globalkeyshortcuts_it.qm`
But IMO the` lxqt2-` needs to be reverted anyway.

**Comment by stefonarch** [comment]
It's not that simple. https://github.com/lxqt/lxqt-build-tools/pull/77/files#r1485031822

**Comment by gfgit** [comment]
> I wasn't following the discussion, but the cause of troubles may have been what I warned about at [lxqt/lxqt-build-tools#77 (comment)](https://github.com/lxqt/lxqt-build-tools/pull/77#discussion_r1247857714). Sadly, @luis-pereira hasn't paid attention to it, and I don't intend to make things more complicated by adding commits to his PR.

Hi, I've pushed new changes needed to build `lxqt-panel`. To build `lxqt-globalkeys` I've first removed the "2" suffix mentioned here (regarding installation directory).

The project name still has the suffix but it can be removed, I'm still confused about where to put it :)

**Comment by tsujan** [comment]
I can just hope that @luis-pereira will hear us.

BTW, I might (and, most probably, will) join you and @stefonarch when a stable version of KF6 comes out. Keep up the good work.

**Comment by stefonarch** [comment]
> > I wasn't following the discussion, but the cause of troubles may have been what I warned about at [lxqt/lxqt-build-tools#77 (comment)](https://github.com/lxqt/lxqt-build-tools/pull/77#discussion_r1247857714). Sadly, @luis-pereira hasn't paid attention to it, and I don't intend to make things more complicated by adding commits to his PR.
> 
> Hi, I've pushed new changes needed to build `lxqt-panel`. To build `lxqt-globalkeys` I've first removed the "2" suffix mentioned here (regarding installation directory).
> 
> The project name still has the suffix but it can be removed, I'm still confused about where to put it :)

I think we don't need any "2" anywhere at the end. On the test installation I'm using just `lxqt-` and as we won't support Qt5 anymore there's no need to distinguish anything IMO, exept eventually the project name `lxqt2-build-tools`.

**Comment by luis-pereira** [comment]
> I can just hope that @luis-pereira will hear us.
> 
> BTW, I might (and, most probably, will) join you and @stefonarch when a stable version of KF6 comes out. Keep up the good work.

@tsujan Yes, I hear. Real life stuff has been on the way.
What needs to be done ?

File: CMakeLists.txt [hunk 1]
```diff
@@ -4,29 +4,29 @@ cmake_minimum_required(VERSION 3.1.0 FATAL_ERROR)
 # The cmake_minimum_required(VERSION) command implicitly invokes the cmake_policy(VERSION)
 # command to specify that the current project code is written for the given range of CMake
 # versions.
-project(lxqt-globalkeys)
+project(lxqt2-globalkeys)
```
Comment (tsujan) [inline CMakeLists.txt:7]: Why changing `lxqt` to `lxqt2` here? `lxqt-globalkeys` is a component that should be replaced, not installed in parallel to its Qt5 version.

Comment (stefonarch) [reply CMakeLists.txt:7]: That's a leftover from the beginning, atm panel looks for this.

Comment (tsujan) [reply CMakeLists.txt:7]: > atm panel looks for this.

Yes, I noticed it in the panel port first. But future reviews will be done from top to bottom.

**Comment by romangg** [comment]
I get a compile error with the current version:

```
57/58] Linking CXX shared library ui/liblxqt2-globalkeys-ui.so.2.0.0
FAILED: ui/liblxqt2-globalkeys-ui.so.2.0.0 
: && /usr/bin/c++ -fPIC -fno-exceptions -Wall -Wextra -Wchar-subscripts -Wno-long-long -Wpointer-arith -Wundef -Wformat-security -Wnon-virtual-dtor -Woverloaded-virtual -Wpedantic -fdiagnostics-color=always -g  -Wl,--no-undefined -Wl,-Bsymbolic-functions -shared -Wl,-soname,liblxqt2-globalkeys-ui.so.2 -o ui/liblxqt2-globalkeys-ui.so.2.0.0 ui/CMakeFiles/lxqt2-globalkeys-ui.dir/lxqt2-globalkeys-ui_autogen/mocs_compilation.cpp.o ui/CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o  -Wl,-rpath,/home/roman/dev/lxqt/lxqt-globalkeys/build/client:  client/liblxqt2-globalkeys.so.2.0.0  /usr/lib/libQt6Widgets.so.6.6.2  /usr/lib/libQt6Gui.so.6.6.2  /usr/lib/libGLX.so  /usr/lib/libOpenGL.so  /usr/lib/libQt6DBus.so.6.6.2  /usr/lib/libQt6Core.so.6.6.2 && :
/usr/bin/ld: ui/CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: warning: relocation against `_ZN17GlobalKeyShortcut6Client18grabShortcutFailedEv' in read-only section `.text'
/usr/bin/ld: ui/CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: in function `ShortcutSelector::ShortcutSelector(QWidget*)':
/home/roman/dev/lxqt/lxqt-globalkeys/source/ui/shortcut_selector.cpp:40:(.text+0x5c): undefined reference to `GlobalKeyShortcut::Client::instance()'
/usr/bin/ld: /home/roman/dev/lxqt/lxqt-globalkeys/source/ui/shortcut_selector.cpp:53:(.text+0x1d2): undefined reference to `GlobalKeyShortcut::Client::grabShortcutCancelled()'
/usr/bin/ld: ui/CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: relocation R_X86_64_PC32 against undefined protected symbol `_ZN17GlobalKeyShortcut6Client21grabShortcutCancelledEv' can not be used when making a shared object
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

I'm not sure why this happens since CMAKE_POSITION_INDEPENDENT_CODE is set in the toplevel directory. I'm tried compiling with gcc and clang.

**Comment by stefonarch** [comment]
With gcc the same, while in Neon it ran fine.
```

[ 81%] Linking CXX shared library liblxqt2-globalkeys-ui.so
/usr/bin/ld: CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: warning: relocation against `_ZN17GlobalKeyShortcut6Client18grabShortcutFailedEv' in read-only section `.text'
/usr/bin/ld: CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: in function `ShortcutSelector::ShortcutSelector(QWidget*)':
/home/stef/git/gfgit/lxqt-globalkeys/ui/shortcut_selector.cpp:40:(.text+0x5c): undefined reference to `GlobalKeyShortcut::Client::instance()'
/usr/bin/ld: /home/stef/git/gfgit/lxqt-globalkeys/ui/shortcut_selector.cpp:53:(.text+0x1d2): undefined reference to `GlobalKeyShortcut::Client::grabShortcutCancelled()'
/usr/bin/ld: CMakeFiles/lxqt2-globalkeys-ui.dir/shortcut_selector.cpp.o: relocation R_X86_64_PC32 against undefined protected symbol `_ZN17GlobalKeyShortcut6Client21grabShortcutCancelledEv' can not be used when making a shared object
/usr/bin/ld: link finale non riuscito: bad value
collect2: error: ld returned 1 exit status

```

File: CMakeLists.txt (hunk 1 above)
Comment (gfgit) [reply CMakeLists.txt]: I've removed every "2" suffix, including renamed translation files. This is now "2" free version.

**Comment by gfgit** [comment]
@romangg @stefonarch I don't have those errors in Neon. It builds fine.
```
CMake 3.28.3
GCC 11.4.0
ldd 2.35
GNU Binutils 2.38
```

**Comment by stefonarch** [comment]
It doesn't build on arch, in VM Neon all is fine.

**Comment by romangg** [comment]
Yea, I'm on Manjaro. 
```
CMake 3.28.3
GCC 13.2.1
ldd 2.39
GNU Binutils 2.42
```

I also tried with Mold linker. This also fails, but the error message is a bit more descriptive:

```
[99/102] Linking CXX shared library ui/liblxqt-globalkeys-ui.so.2.0.0
FAILED: ui/liblxqt-globalkeys-ui.so.2.0.0 
: && /usr/bin/c++ -fPIC -fno-exceptions -Wall -Wextra -Wchar-subscripts -Wno-long-long -Wpointer-arith -Wundef -Wformat-security -Wnon-virtual-dtor -Woverloaded-virtual -Wpedantic -fdiagnostics-color=always -O3 -DNDEBUG  -Wl,--no-undefined -Wl,-Bsymbolic-functions "-fuse-ld=mold" -shared -Wl,-soname,liblxqt-globalkeys-ui.so.2 -o ui/liblxqt-globalkeys-ui.so.2.0.0 ui/CMakeFiles/lxqt-globalkeys-ui.dir/lxqt-globalkeys-ui_autogen/mocs_compilation.cpp.o ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o  -Wl,-rpath,/home/roman/dev/lxqt/lxqt-globalkeys/build/client:  client/liblxqt-globalkeys.so.2.0.0  /usr/lib/libQt6Widgets.so.6.6.2  /usr/lib/libQt6Gui.so.6.6.2  /usr/lib/libGLX.so  /usr/lib/libOpenGL.so  /usr/lib/libQt6DBus.so.6.6.2  /usr/lib/libQt6Core.so.6.6.2 && :
mold: error: undefined symbol: GlobalKeyShortcut::Client::grabShortcutFailedv
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::instancev
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::grabShortcutCancelledv
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::grabShortcutTimedoutv
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::shortcutGrabbedRK7QString
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::staticMetaObject
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::ShortcutSelector(QWidget*))
mold: error: undefined symbol: GlobalKeyShortcut::Client::cancelShortcutGrabv
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::grabShortcuti)>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::cancelNowv)>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::cancelNowv)
mold: error: undefined symbol: GlobalKeyShortcut::Client::grabShortcutj
>>> referenced by shortcut_selector.cpp
>>>               ui/CMakeFiles/lxqt-globalkeys-ui.dir/shortcut_selector.cpp.o:(ShortcutSelector::grabShortcuti)
collect2: error: ld returned 1 exit status
[100/102] Building CXX object daemon/CMakeFiles/lxqt-globalkeysd.dir/core.cpp.o
ninja: build stopped: subcommand failed.
```

**Comment by romangg** [comment]
This fixes it for me:
```diff
diff --git a/ui/CMakeLists.txt b/ui/CMakeLists.txt
index 6388911..525b4f9 100644
--- a/ui/CMakeLists.txt
+++ b/ui/CMakeLists.txt
@@ -147,11 +147,6 @@ target_include_directories(${PROJECT_NAME}
     INTERFACE "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/include/${PROJECT_NAME}>"
 )
 
-target_compile_definitions(${PROJECT_NAME}
-    PRIVATE
-        "SHARED_EXPORT=Q_DECL_EXPORT"
-)
-
 export(TARGETS ${PROJECT_NAME} APPEND FILE "${CMAKE_BINARY_DIR}/${PROJECT_NAME}-targets.cmake")
```

**Comment by romangg** [comment]
From the docs it says that `Q_DECL_EXPORT` should be used when exporting and `Q_DECL_IMPORT` when importing stuff. When `SHARED_EXPORT` is not defined it defaults to the second. So it makes sense that in the importing binary it should be removed.

But from my experience a separate export marking was only ever necessary in the exporting header, not importing ones, so I don't know what `Q_DECL_IMPORT` actually does.

**Comment by romangg** [comment]
Correction: the ui subdirectory is also defining a library. So we should export there something too.

**Comment by romangg** [comment]
Your latest commit works well with the two different names. :+1:

**Comment by gfgit** [comment]
I will check other libraries. Macros in public headers should always have unique names

**Comment by gfgit** [comment]
So far other libraries involved in panel seem to have unique export macro name

File: client/client.cpp [hunk 2]
```diff
@@ -123,7 +124,9 @@ void ClientImpl::registrationFinished(QDBusPendingCallWatcher *watcher)
 
 Action *ClientImpl::addClientAction(const QString &shortcut, const QString &path, const QString &description, QObject *parent)
 {
-    if (!QRegExp(QStringLiteral("(/[A-Za-z0-9_]+){2,}")).exactMatch(path))
+    // FIXME: check if QRegExp syntax is the same
+    static QRegularExpression regexp(QStringLiteral("(/[A-Za-z0-9_]+){2,}"));
+    if (!regexp.match(path).hasMatch())
```
Comment (tsujan) [inline client/client.cpp:129]: @gfgit 
`QRegularExpression::anchoredPattern` should be used for exact matching. After using it, please remove the "FIXME" comment.

**Comment by tsujan** [comment]
@gfgit 
BTW, it could be of great help if you make your branches inside LXQt itself in future. Most of the time, LXQt devs don't make changes to each other's branches without explicit permission, but exceptional cases like [this](https://github.com/lxqt/lxqt-globalkeys/pull/278#discussion_r1533604282) could be handled quickly then.

File: client/client.cpp (hunk 2 above)
Comment (stefonarch) [reply client/client.cpp:129]: How about merge it and fix it afterwards?
It's the last piece missing for having all qt6-ports in git and we need more testers IMO.

Comment (gfgit) [reply client/client.cpp:129]: I can fix it today

Comment (gfgit) [reply client/client.cpp:128]: Done. What about duplicated `joinCommandLine()` method?

**Comment by gfgit** [comment]
> @gfgit BTW, it could be of great help if you make your branches inside LXQt itself in future. Most of the time, LXQt devs don't make changes to each other's branches without explicit permission, but exceptional cases like [this](https://github.com/lxqt/lxqt-globalkeys/pull/278#discussion_r1533604282) could be handled quickly then.

Ok. Should I move my existing open PRs to this repo?

**Review by tsujan — APPROVED** [review]

**Comment by tsujan** [comment]
> Ok. Should I move my existing open PRs to this repo?

That'll be much better, especially now that two of your PRs have been merged.

**Pull request merged**